Use multipledispatch for handlers in assumptions module

JSS95 opened the pull request:
<!-- Your title above should be a short description of what
was changed. Do not include the issue number in the title. -->

#### References to other Issues or PRs
<!-- If this pull request fixes an issue, write "Fixes #NNNN" in that exact
format, e.g. "Fixes #1234" (see
https://tinyurl.com/auto-closing for more information). Also, please
write a comment on that issue linking back to this pull request once it is
open. -->
Fixes #20209 

#### Brief description of what is fixed or changed

Assumption handlers are now multipledispatch instance. According to this, `Predicate` now use a single handler instead of a list of handlers. Also, `remove_handler` function is removed, and parameter type for `register_handler` is changed.

#### Other comments

This PR will break backwards compatibility for user-defined predicates.

#### Release Notes

<!-- Write the release notes for this release below. See
https://github.com/sympy/sympy/wiki/Writing-Release-Notes for more information
on how to write release notes. The bot will check your release notes
automatically to see if they are formatted correctly. -->

<!-- BEGIN RELEASE NOTES -->

- assumptions
    - `Predicate` now use a single handler instead of a list of handlers.
    - Handlers are now multipledispatch instance. This breaks backwards compatibility.
    - `remove_handler` function is removed, and parameter type for `register_handler` is changed. This breaks backwards compatibility.

<!-- END RELEASE NOTES -->

JSS95 commented:
@asmeurer 

This is still in progress, so please take a look at the overall idea.

Previous:
- `Ask[...]Handler` class itself was used as handler object for `Q.[...]` key.

Now:
- `Ask[...]Handler` is now instance of `Ask[...]HandlerClass`, which is subclass of `Dispatcher`.
- To allow inheritance of dispatched signatures, every dispatching is done in `__init__` method. User can dispatch custom handler normally.

JSS95 commented:
Also, I think we need separate classes for each predicates. For example, what `Q.even` returns have to be an instance of `EvenPredicate` (which is subclass of `Predicate`), not an instance of bare `Predicate`.

JSS95 commented:
Done. Let me give a brief summary.

1. Handler is now instance of handler class, which is subclass of `Dispatcher`.
2. Every predicate now has its own class and own handler. For this, `assumptions/predicates` directory is added.
3. Handlers are no longer in `args` of the predicate. This makes `Predicate` immutable.

For example, `Q.even` returns instance of `EvenPredicate`, which is defined in `predicates/ntheory`. `EvenPredicate` has a property `handler`, which returns instance of `AskEvenHandler`. It is a multipledispatch, and defined in `handlers/ntheory`.

Please review this. After all is approved, I will start writing the document.

JSS95 commented:
Most of the changes are just code migrations. I suggest reviewing `handlers/common.py`, `predicates/predicate.py`, and `predicates/common.py` first.

JSS95 commented:
@oscarbenjamin @eric-wieser 
Can you review this, please? Tests have passed (Travis failure is because of documentation error) but I think overall design can be improved.

oscarbenjamin commented on sympy/assumptions/handlers/matrices.py:
```diff
@@ -228,46 +273,54 @@ def MatPow(expr, assumptions):
+        for sig in (
+        MatMul, MatPow, MatrixSymbol, MatrixSlice
+        ):
+            self.register(sig)(getattr(self, sig.__name__))
```
This seems to be a common idiom here. Is there a better way of doing this?

I do wonder whether it would make more sense to define these handlers in the matrices modue rather than here in assumptions.

oscarbenjamin replied:
Perhaps there could be a decorator to indicate which classes a handler method should be registered with like:
```python
    @register(MatPow, MatMul):
    def MatMul():
        ....
```

JSS95 replied:
Decorator seems better. I will try this way.

> I do wonder whether it would make more sense to define these handlers in the matrices modue rather than here in assumptions.

IMO defining in assumtions/handlers makes them more easy to be found.

JSS95 replied:
I modified this in other way. Instead of keeping the classes, I made multipledispatch handlers copyable. Now, for example, `AskExtendedRealHandler` and `AskRealHandler` are both instances of `AskHandlerClass` instead of having each one's own class. It is just that former is copied from the latter and dispatched with some more functions.

oscarbenjamin commented:
I'm not sure I quite follow what the design here does. Note that I wasn't really familiar with the new assumptions code before these changes. Can you explain more clearly what the different classes are and what is being changed either here or in the issue?

The overall diff is large which makes it hard to see what is actually being changed.

asmeurer commented:
Do we need to worry about backwards compatibility here? Is it possible currently for end-users to define their own predicate handlers?

JSS95 commented:
@asmeurer 
Defining custom predicate is possible. See `doc/src/modules/assumptions/index.rst`. However, this breaks backward compatibility.

asmeurer commented:
I'm not sure about the new class design here. It seems like there is a lot more redundancy and boilerplate compared to the way it currently is.

JSS95 commented:
@asmeurer 
Then perhaps I can shave off some parts. Which, in particular, do you think that it's redundancy?
1. That handler is instance of multipledispatch class
2. That every predicate has its own class
3. That user needs more work to define custom predicate

eric-wieser commented on sympy/assumptions/handlers/calculus.py:
```diff
@@ -52,7 +63,7 @@ def Symbol(expr, assumptions):
 
     @staticmethod
-    def Add(expr, assumptions):
+    def Add(expr, *, assumptions):
```
Changing assumptions to be keyword-only is making this diff huge, and as far as I can tell is orthogonal. For the sake of making review easier, can you either:

* Make a PR that does nothing other than make assumptions keyword-only, and rebase this PR once that gets merged
* Remove these changes from this PR, and make the keyword-only arguments a follow-up

Thanks!

eric-wieser commented on sympy/assumptions/ask.py:
```diff
@@ -25,1124 +26,213 @@ class AssumptionKeys:
-        # TODO: Add examples
-        return Predicate('hermitian')
+        from .predicates.sets import HermitianPredicate
+        return HermitianPredicate('hermitian')
```
Passing string names into these classes feels useless to me - would anyone ever use `HermitianPredicate('not hermitian')`?

eric-wieser commented on sympy/assumptions/ask.py:
```diff
@@ -25,1124 +26,213 @@ class AssumptionKeys:
-        """
-        Hermitian predicate.
-
-        ``ask(Q.hermitian(x))`` is true iff ``x`` belongs to the set of
-        Hermitian operators.
-
-        References
-        ==========
-
-        .. [1] http://mathworld.wolfram.com/HermitianOperator.html
-
```
I like the way you have moved these docstrings - it means that `help(Q.hermitian)` gives something useful, which currently it doesn't.

asmeurer commented:
We should consider whether it is actually ideal for predicates to know about their handlers. The idea behind multipledispatch is that dispatched functions are independent of the objects they define, as opposed to usual OO where functions are defined as methods on objects. I think it would be a simpler design to have the handlers be completely separate. It also makes more sense to me for the assumptions predicates to be separate from the facts about them in the code (as I expressed [here](https://github.com/sympy/sympy/pull/20148#issuecomment-703887787)). 

This is already breaking compatibility with the way handlers work on predicates (e.g., `add_handler` is gone), so if we are committed to doing that, then I think we should try to make a good design.

asmeurer commented:
I don't like that this breaks `Predicate("name")` as a way to create predicates. Subclassing is fine, but we should try to make the old way work. Actually, with something like https://github.com/sympy/sympy/pull/2538, we can continue to instantiate predicates in the old way and include documentation for them. Is there a reason to make Predicate a superclass instead of an instantiator, other than to include docstrings and handler methods?

asmeurer commented:
I guess an advantage is that the class name is different for each predicate. But I'm not sure if that is actually something that we need. Predicates are already Basic instances, so there's no need to use `isinstance` to compare them. It also means for any custom predicate you have to come up with a class name in addition to the predicate name. Another issue is if we did want to continue to support something like `Predicate("custom")`, we would have to dynamically create a class, similar to `Function('f')`. But I think this is a bad design that we should avoid copying. I much prefer the Predicate design where the unapplied class is Basic. It avoids lots of problems that come from dynamically creating functions, such as issues with metaclasses, problems with pickling, and general confusion from almost everyone as to how the class heirarchy actually works.

This is the source of most of the redundancy and boilerplate I mentioned earlier.

JSS95 commented:
@asmeurer 
I reverted some changes. Predicate subclasses are gone, and `Predicate("name")` is supported now.

JSS95 commented:
Done. Please review.  
Most of the diff comes from `assumptions/handlers`. Please take a look at the handlers in `handlers/common` and `AskExtendedRealHandler` in `handlers/sets` to see how handlers are copied and extended to generate sub-handler.

JSS95 commented:
No review have been made for a week, so I assume that this commit is getting unanimous approval. I'll start writing the documentation.

JSS95 commented:
@asmeurer @oscarbenjamin @eric-wieser Can this be merged?

JSS95 commented:
@asmeurer @oscarbenjamin @eric-wieser It's finished. Please merge this.

JSS95 committed:
Implement multipledispatch handler

JSS95 commented:
Squashed the commits

oscarbenjamin commented:
I'm sorry I haven't had time to look at this. I'm still struggling to get the sympy 1.7 release out. One thing after another keeps going wrong

JSS95 commented:
Oh, sorry. Didn't know that 1.7 is coming out. No need to rush, then!

JSS95 commented:
@oscarbenjamin Or do you thing this can be added to 1.7 release?

oscarbenjamin commented:
I'm not sure I understand all of the changes here yet but I don't see a need to rush this into the release. Big changes are best made early in the release cycle.

JSS95 commented:
OK, I agree.

JSS95 committed:
Fix typo and enhance docstring of register_handler

JSS95 committed:
Enhance Predicate's default dispatcher

When Predicate with no handler is generated, a copied
CommonHandler is used as handler instead of Dispatcher.

JSS95 commented:
@asmeurer @eric-wieser Can you spare some time to take a look on this? New design is explained in [this comment](https://github.com/sympy/sympy/pull/20215#issuecomment-708111437), and also in the documentation in this commit.

czgdp1807 commented:
@JSS95 Please resolve the conflicts. Thanks for your contributions.

czgdp1807 commented:
IMHO, this PR is too big to be reviewed efficiently. Could you please break into smaller parts?

eric-wieser commented on sympy/assumptions/handlers/matrices.py:
```diff
@@ -6,697 +6,769 @@
+### AskSquareHandler ###
 
-    Inverse = Transpose
+AskSquareHandler = CommonHandler.copy(
+    'AskSquareHandler',
+    doc="""
+    Handler for key 'square'
+    """
+)
```
I don't understand why we need all these objects - why can't we use `Q.square.handler.register` instead of `AskSquareHandler.register`? Is it an import issue? A lifetime issue?

JSS95 replied:
We can use `Q.square.handler.register` if `AskSquareHandler` is registered to `Q.square`. I updated the document - please see `ask.register_handler`.

I implemented `copy` method because `AskSquareHandler` must inherit every function dispatched to `CommonHandler`. Previously, these two were classes and inheritance was done by subclassing. But in this design, these two are multipledispatch instance so I had to make this change.

If you have any other idea please let me know. Thanks!

eric-wieser replied:
What handlers does `CommonHandler` have that we actually care about? Are there any cases where there are multiple levels of base classes, `SomeHandler -> Other handler -> CommonHandler`? Or is it always just one level?

JSS95 replied:
In  `assumptions.handlers.sets`, we have `AskExtendedRealHandler`  which is derived from `AskRealHandler`. It's `AskExtendedRealHAndler -> AskRealHandler -> CommonHandler`.

eric-wieser replied:
Could this be handled by registering a dispatcher default that calls the "base" handler?

JSS95 replied:
Perhaps I can add `fallback_dispatcher` attribute to `AskHandlerClass`. I will try that way and report here.

JSS95 replied:
I removed `copy` method and added `base` parameter to `AskHandlerClass.__init__` in commit `f3247ab`. Now, the syntax is
```python
AskSquareHandler = HandlerClass('AskSquareHandler', doc='...', base=CommonHandler)
```
Here, `AskSquareHandler` inherits every functions dispatched to `CommonHandler`.

JSS95 replied:
@eric-wieser, if you have any other comment, please leave your review in #20475. Thanks!

JSS95 committed:
Update test_query.test_custom_AskHandler

The test for MersenneHandler was following the older commit's design

JSS95 committed:
Merge branch 'master' into assumptions_handlers

JSS95 committed:
Improve code quality

JSS95 committed:
Update doctest of register_handler

eric-wieser commented:
This isn't actually multiple dispatch, right? It's just single dispatch. Given that, how does `functools.singledispatch` compare in performance and convenience to `Dispatcher`?

JSS95 commented:
@eric-wieser 
This will be expanded to multipledispatch handlers in succeeding PRs. Currently, SymPy has only unary predicates such as `Q.even`, `Q.square`, etc. I am trying to implement polyadic predicates such as `Q.coprime`. With this, handlers need to be dispatched with multiple arguments. Please refer to #20153.

JSS95 commented:
@czgdp1807 I will make some enhancements here and try that way.

JSS95 committed:
Change the inheritance of handlers

AskHandlerClass now take `base` parameter.

JSS95 commented:
I broke this down to separate PRs. @czgdp1807, please see #20475. I am closing this PR.
